sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -1408,11 +1409,37 @@ case class ValidateExternalType(child: Expression, expected: DataType)
 
   override def dataType: DataType = RowEncoder.externalDataTypeForInput(expected)
 
-  override def eval(input: InternalRow): Any =
-    throw new UnsupportedOperationException("Only code-generated evaluation is supported")
-
   private val errMsg = s" is not a valid external type for schema of ${expected.simpleString}"
 
+  private lazy val checkType = expected match {
+    case _: DecimalType =>
+      (value: Any) => {
+        Seq(classOf[java.math.BigDecimal], classOf[scala.math.BigDecimal], classOf[Decimal])
+          .exists { x => value.getClass.isAssignableFrom(x) }
+      }
+    case _: ArrayType =>
+      (value: Any) => {
+        value.getClass.isAssignableFrom(classOf[Seq[_]]) || value.getClass.isArray
```

Comment:
What is cheaper? `isAssignableFrom ` or `isArray`? If it is the latter then we should swap the order.

Reply:
I have no idea for that and what do u know about it in jvm? @kiszk @rednaxelafx

Reply:
Hi guys, sorry I'm late.

In your new code you're doing:
```diff
+    case _: ArrayType =>
+      (value: Any) => {
+        value.getClass.isArray || value.isInstanceOf[Seq[_]]
+      }
```
which is good. `xxx.getClass().isAssignableFrom(some_class_literal)` in the old version of this PR is actually backwards, it should have been `some_class_literal.isAssignableFrom(xxx.getClass())`, e.g.
```
scala> classOf[String].isAssignableFrom(classOf[Object])
res0: Boolean = false

scala> classOf[Object].isAssignableFrom(classOf[String])
res1: Boolean = true
```
and the latter is semantically the same as `xxx.isInstanceOf[some_class]`. `isInstanceOf[]` is guaranteed to be at least as fast as `some_class_literal.isAssignableFrom(xxx.getClass())`, and in general `isInstanceOf[]` is faster.

`xxx.getClass().isArray()` has a fixed overhead, whereas `isInstanceOf[]` can have a fast path slightly faster than the `isArray` and a slow path that can be much slower than `isArray`. So putting the `isArray` check first in your new code makes more sense to me.

Reply:
For those curious:

In HotSpot, the straightforward interpreter/C1 implementation of `xxx.getClass().isArray()` path is actually something like:
```
// for getClass()
klazz = xxx._klass; // read the hidden klass pointer field from the object header
clazz = klazz._java_mirror; // read the java.lang.Class reference from the Klass
// for clazz.isArray(): go through JNI and call the native JVM_IsArrayClass() inside HotSpot
klazz1 = clazz->_klass;
result = klazz1->oop_is_array();
```
So a JNI native method call is involved and that's not really fast. But C2 will optimize this into something similar to:
```
klazz = xxx._klass;
result = inlined klazz->oop_is_array();
```
So that's pretty fast. No need to load the `java.lang.Class`  (aka "Java Mirroe") reference anymore.

In the `xxx.isInstanceOf[Seq[_]]` case, again the interpreter version would go through a JNI native method call, whereas the C1/C2 versions will inline a fast path logic and do a quick comparison against a per-type cache. This fast path check has similar overhead to the C2 `isArray()` overhead.